README.md
# Secure-Audit-Log-Notarizer

_Audit log protection mechanism for popHealth certification - d(3) criteria_

This module is designed to reveal tampering in database entries. It was designed to accompany PopHealth to protect its audit log.

The idea is that after a break-in to the host computer, or any suspicious action in the database, we need to check to see if information has been tampered with.

## Architecture

The server component, ````routes/notar-routes.js```` uses ````app.js```` to offer itself as a web service. 

The client component ````lib/client.js```` reads a MongoDB collection, hashes each entry and sends the hash and item ID to the server. 

The server enters the ID and the hash into its database and returns the 'notarizing' key to the entry. 

When it is time to check an entry, the client goes through its secured collection, and for each item, sends its hash, ID and the 'notarizing' key. The server checks its database and flags any invalid entries, both in its database and in the return message to the client.

Optionally, the client can also send a ZIP of each item to the server and the server will put those into a big zip file, starting a new zip file when the configured size limit is reached.

## Installation

After cloning from Github, cd into the application directory and run:

````npm install````

If you want to run the server and/or client as a daemon (recommended), also install ````forever````:

````npm install -g forever````

## Usage

For development, server and client can be run on the same host. In production, it is essential that the server run on another host in the LAN. An SSL option is provided for that purpose.

### Server

````node routes/notar-routes.js [--port PORT] [ --ssl ] [--level LOGLEVEL] [--serverdb SERVER_DATABASE ]````

Or as demon:

   ````forever start routes/notar-routes.js [--port PORT] [ --ssl ] [--level LOGLEVEL] [--serverdb SERVER_DATABASE ]````

(See also *Configuration* section.)

### Client

To start monitoring client:

````node lib/client.js [--host NOTARIZER_SERVER] [--ssl] [--config CONFIGPATH] [--level LOGLEVEL]````

Or as daemon:

````forever start lib/client.js [--host NOTARIZER_SERVER] [--ssl] [--config CONFIGPATH] [--level LOGLEVEL]````

### Validation Client

To run validation of the notarizer database collection entries: 

````node lib/client.js --validate [ OPTIONAL_ARGS ]````

## Configuration

There are a number of configuration options in config/config.js. These can be modified to tailor the application to your environment and to avoid the need for any command-line arguments.

### Config items

* MongoDB authentication. If your MongoDB runs with --auth, you must create a credentials.js file in the config/ directory:

    ````echo 'module.exports="USER:PASSWORD@"' > config/credentials.js````

    If your MongoDB installation runs without authentication, then comment out the ````require('./credentials')```` line and remove ````'+cred+'```` from the ````mongodb.urlskel```` entry in config.js

* certs: if running under SSL, you must supply (PEM) certificates and keys here. They may be self-signed (caCert helps with this). 

    (If you have self-signed certificates and test with curl, use ````-k```` to accept the server certificate because curl does not have any provision for self-signed certificates. Also use ````--key CLIENTKEY --cert CLIENTCERT```` to supply client-certificates.)

* ssl: if this is true, then certs must be supplied.

* port: port the server will listen to.

* mongodb: Both client and server information are here (and can be changed independently)

  * mongodb.logs: database collection to be monitored
  * mongodb.nota: server collection where hashes, etc., will be stored
  * mongodb.since: earliest date in logs to consider.

* notarizer: information the client uses to access the server. You will change the ````notarizer.host```` for production use

* zip: configuration for the option to store zipped entries on the server as a backup.

* https: most values are filled in dynamically. 
  
  * https.requestCert : whether client must have a client certificate to access the server.
  * https.rejectUnauthorized: true for production, false if you want to debug authorization. False requires an ````app.use()```` clause which checks ````req.client.authorized```` and perhaps outputs an error message if client is not authorized.

* routes: currently ignored
* requestOpts: for the client; passphrase would be the passphrase on the client key/cert if any.

* sha2: the hash function used to hash the log entries.



CHANGELOG
# 0.0.1
  Manual end-to-end tests run against a PopHealth database.

	


test/testZip.js
/** testZip.js
 * Copyright 2017 Zato, Inc.
 *
 * Author: jbroglio
 * Date: 2/26/17
 * Time: 9:52 AM
 */
var fs=require('fs'),
    Zipper=require('jszip'),
    path=require('path'),
    util=require('util'),
    bunyan=require('bunyan'),
    log=bunyan.createLogger({name:'testZip', level:'DEBUG'})
;

process.on('uncaughtException',function(err){
  log.error(err, "Zipping");
  if (err.stack) util.format("%j", err.stack);

})


var zipMem=new Zipper();
var newzip=new Zipper();
zipMem.file("log1.txt", "This is log one");
zipMem.generateNodeStream({type:'nodebuffer',streamFiles:true})
          .pipe(fs.createWriteStream('zipmem.zip'))
          .on('finish', function () {
            // JSZip generates a readable stream with a "end" event,
            // but is piped here in a writable stream which emits a "finish" event.
            console.log("out.zip written.");
            var data ="";
            fs.createReadStream('zipmem.zip','binary')
                .on('data', function (chunk) {
                  data+=chunk;
                })
                .on('close', function () {
                  newzip.loadAsync(data)
                      .then(function (z) {
                        z.file('log2.txt', "This is log 2");
                        z.generateNodeStream({type: 'nodebuffer', streamFiles: true})
                            .pipe(fs.createWriteStream('ziptestb.zip'))
                            .on('finish', function () {
                              // JSZip generates a readable stream with a "end" event,
                              // but is piped here in a writable stream which emits a "finish" event.
                              console.log("out.zip 2 written.");
                            });
                      })
                      .catch(function (err) {
                        log.error(err)
                      })
                })
          })





app.js
var bunyan=require('bunyan')

var express = require('express');
var path = require('path');
var favicon = require('serve-favicon');
var logger = require('morgan');
var cookieParser = require('cookie-parser');
var bodyParser = require('body-parser');

//var index = require('./routes/index');
//var users = require('./routes/users');

var app = express();

// view engine setup
app.set('views', path.join(__dirname, 'views'));
//app.set('view engine', 'jade');

// uncomment after placing your favicon in /public
//app.use(favicon(path.join(__dirname, 'public', 'favicon.ico')));
//app.use(logger('dev'));
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));
app.use(cookieParser());
app.use(express.static(path.join(__dirname, 'public')));

//app.use('/', index);
//app.use('/users', users);

// catch 404 and forward to error handler

// error handler
app.use(function(err, req, res, next) {
  // set locals, only providing error in development
  res.locals.message = err.message;
  res.locals.error = req.app.get('env') === 'development' ? err : {};

  // render the error page
  res.status(err.status || 500);
  res.render('error');
});

module.exports = app;


lib/client.js
/** client.js
 * Copyright 2017 Zato, Inc.
 *
 * Author: jbroglio
 * Date: 2/24/17
 * Time: 7:59 AM
 */
var bunyan = require('bunyan'),
    log = bunyan.createLogger({name: 'client', level: 'INFO'}),
    util = require('util'),
    path = require('path'),
    request = require('request-promise'),
    Zipper = null,
    argv = require('yargs')
        .usage('node client.js [--host NOTARIZER_SERVER] [--ssl] [--config CONFIGPATH] [--level LOGLEVEL] [--validate]')
        .boolean('ssl')
        .argv,
    mongoskin = require('mongoskin'),
    extend = require('node.extend'),
    configPath = path.resolve(__dirname, '../config/')
    ;

var config, options, db, logsdb, notadb;
var cursorOptions = {
  tailable: true,
  awaitdata: true,
  numberOfRetries: -1
};

var username;
var password;
function getSignature(hash, obid, txt, cb) {
  if (typeof txt === 'function') {
    cb = txt;
    txt = null;
  }
  var opts = extend({
        body: {hash: hash, id: obid},
        method: "POST",
        uri: config.notarizer.url + '/sign'
      },
      options);
  if (txt) {
    if (config.zip.active && !Zipper) Zipper = require('jszip');
    var zipped = new Zipper();
    zipped.file(obid, txt);
    zipped.generateAsync({type: 'string'})
        .then(function (s) {
          opts.body.zipped = s;
          sendIt()
        })
        .catch(function(err){
          log.error(err," generateAsync zip");
        })
  } else sendIt();

  function sendIt() {
    request(opts)
        .then(function (body) {
          return cb(null, body);
        })
        .catch(function (err) {
          log.error(err, "posting log message: %s", obid);
          return cb(err, null)
        })
  }
}

// # validateLogs
// Begins by comparing notarized count with Notarizer's count.
// We do this because log file could keep growing while we are working.
// Then we validate each notarized item.
function validateLogs() {
  notadb.findOne({}, {sort: [["time", "asc"]]}, function (err, doc) {
    if (err) {
      return log.error(err, "querying for validation count start")
    }
    var start = doc.time;
    var startId = doc._id;
    notadb.findOne({}, {sort: [["time", "desc"]]}, function (err, doc2) {
      if (err) {
        return log.error(err, "querying for validation count end")
      }
      var end = doc2.time;
      var endId = doc._id;
      var opts = extend({
        uri: config.notarizer.url + '/count',
        qs: {start: start, end: end},
        method: 'GET'
      }, options);
      request(opts)
          .then(function (result) {
            var remoteCount = result.count;
            var remoteTotal = result.total;
            notadb.count({$and: [{time: {$gte: start}}, {time: {$lte: end}}]}, function (err, count) {
              if (err) {
                return log.error(err, "querying log item count")
              }
              if (remoteCount > count) {
                log.error("VALIDATION FAILED: missing notarized items")
              }
              if (remoteTotal != count) {
                log.warn("notarizedTotal:%d != localTotal: %d", remoteTotal, count)
              }
              if (count==remoteCount) log.info("Validating: Local count:%s == remote count:%s", count, remoteCount);
              //
              var skip1 = remoteTotal - count;
              // validate the log entries
              logsdb.findOne({_id: startId}, function (err, doc1) {
                if (err) {
                  return log.error(err, "VALIDATION FAILED while getting 1st log item by id")
                }
                var logStart = doc1.created_at;
                logsdb.findOne({_id: endId}, function (err, doc2) {
                  if (err) {
                    return log.error(err, "VALIDATION FAILED while getting last log item by id")
                  }
                  var logEnd = doc2.created_at;
                  // now validate all entries. By running from notadb, we will verify that every log item is present
                  validateEntries(start, end, function(err, numberValidated){
                    if (err) log.error(err, "validating");
                    else log.info("Number Validated: "+numberValidated);
                    // exit
                    db.close();
                  });
                })
              })
            })
          })
          .catch(function (err) {
            return log.error(err, "requesting validation count")
          })
    })
  })
}
var numberValidated=0;
function validateEntries(start, end, cb) {
  var skip = 0, limit = 100;
  validateSome();

  function validateSome() {
    notadb.find({$and: [{time: {$gte: start}}, {time: {$lte: end}}]}).skip(skip).limit(limit).toArray()
        .then(function (keyDocs) {
          if (keyDocs) skip += keyDocs.length;
          else {
            return cb(null, numberValidated);
          }
          var query = [];
          getHash()

          function getHash() {
            if (!keyDocs || !keyDocs.length) {
              if (query.length) {
                var opts = extend(
                    {
                      uri: config.notarizer.url + "/validate",
                      body: query,
                      method: 'POST'
                    }, options);
                return request(opts)
                    .then(function (result) {
                      if (result.failed)log.error("VALIDATION FAILURE: %j", result.failed);
                      else numberValidated=result.numberValidated; // this is cumulative on server for now
                      setImmediate(validateSome);
                    })
                    .catch(function (err) {
                      log.error(err, "Calling notarizer with validation batch. Check server availability.");
                      cb(err,numberValidated);
                    })
              } else return cb(null, numberValidated);
            }
            var kd = keyDocs.shift();
            logsdb.findOne({_id: kd._id}, function (err, logitem) {
              if (err) {
                log.error(err, "VALIDATION ERROR. NO log item for nota entry:%s.", kd._id.toString());
                // keep going to list all the errors
              } else {
                var hash = config.sha2(JSON.stringify(logitem));
                query.push({signature: kd.signature, itemid: kd._id.toString(), hash: hash})
              }
              setImmediate(getHash)
            })
          }
        })
        .catch(function(err){
          log.error(err, "validating");
        })
  }
}

function notarizeLogs() {
  var q;
  notadb.findOne({signature: {$ne: "ERROR"}}, {sort: [["time", "desc"]]}, function (err, lastnota) {
    if (err || !lastnota) {
      if (err) log.error(err, "finding last nota doc");
      q = {};
      doNotarize();
    } else {
      var arr = logsdb.findOne({_id: lastnota._id}, function (err, lastLog) {
        if (err) {
          log.error(err, "finding last logs doc");
          q = {};
        } else q = {created_at: {$gt: lastLog.created_at}};
        doNotarize();
      })
    }
  })

  var cursor;

  function cursorNext() {
    if (!cursor) return;
    try {
      cursor.nextObject(function (err, doc) {
        if (err) {
          log.error(err, "on cursor.nextObject()--continuing...")
          return setImmediate(notarizeLogs)
        }
        if (!doc) return setImmediate(notarizeLogs);
        var docString = JSON.stringify(doc);
        var hash = config.sha2(docString);
        getSignature(hash, doc._id.toString(), docString, function (err, val) {
          if (err) {
            return notadb.insert({_id: doc._id, signature: "ERROR"}, function (err, ret) {
              if (err) {
                log.error(err, "insertion error for %s", doc._id.toString())
                return setImmediate(cursorNext)
              }
            })
          }

          notadb.findOne({_id: doc._id, signature: "ERROR"}, function (err, ret) {
            // check if we had an error before
            if (err || !ret) {
              notadb.insert({_id: doc._id, signature: val.signature, time: val.time}, function (err, ret) {
                if (err) {
                  log.error(err, "insertion error for %s", doc._id.toString())
                  return setImmediate(cursorNext);
                }
                log.info("Notarized: %s", doc._id);
                config.mongodb.since = doc.created_at;
                return setImmediate(cursorNext);
              })
            } else {
              notadb.update({_id: doc._id}, {$set: {signature: val.signature, time: val.time}},
                  function (err, ret) {
                    if (err) {
                      log.error(err, "insertion error for %s", doc._id.toString())
                      return setImmediate(cursorNext);
                    }
                    log.info("Notarized: %s", doc._id);
                    // NOW we can update for last time seen
                    config.mongodb.since = doc.created_at;
                    return setImmediate(cursorNext);
                  })
            }
          });
        })
      })
    } catch (err) {
      log.error(err, "Cursor error");
      setImmediate(notarizeLogs());
    }
  }

  function doNotarize() {
    cursor = logsdb.find(q, {tailable: true});
    if (!cursor) {
      return log.error(err, "error tailing logs");
    }

    cursorNext();
  }
}

//================  MAIN ======================
if (require.main == module) {
  var configFile = argv.config || path.resolve(configPath, 'config.js');
  if (argv.clientdb) config.mongodb.clientdb = argv.clientdb;
  if (argv.ssl) config.ssl = true;
  if (argv.host) config.notarizer.host = argv.host;
  if (argv.level) log.level(argv.level);
  config = require(configFile).init();
  options = extend({}, config.requestOpts);
  options.json = true;
  options.jar = true;
  try {
    db = mongoskin.MongoClient.connect(
        util.format(config.mongodb.urlskel, config.mongodb.clientdb),
        {native_parser: true});
  } catch (err) {
    dbConnectFailure(err);
  }

  logsdb = db.collection(config.mongodb.logs);
  notadb = db.collection(config.mongodb.nota);
  config.notarizer.url = util.format("%s://%s:%d/%s",
      (config.ssl ? "https" : "http"),
      config.notarizer.host, config.notarizer.port, config.notarizer.app);

  if (argv.validate)
    validateLogs();
  else
    notarizeLogs();

  function dbConnectFailure(err) {
    var loc = config.mongodb.urlskel.indexOf('@') + 1
    log.error(err, 'Unable to connect to mongo at %s.\n   ', config.mongodb.urlskel.substring(loc));
    cleanup();
    process.exit(1);
  }

}

config/config.js
/** config.js
 * Copyright 2017 Zato, Inc.
 *
 * Author: jbroglio
 * Date: 2/24/17
 * Time: 7:58 AM
 */
var path = require('path'),
    certPath = path.resolve(__dirname, 'cert'),
    fs = require('fs'),
    util = require('util'),
    bunyan=require('bunyan'),
    log = bunyan.createLogger({name: 'config'}),
    extend = require('node.extend'),
    crypto = require('crypto'),
    cred = require('./credentials')
    ;

// overrides is another config with just the changes filled in.

var config =
{
  site: '',
  certs: {
    clientCert: path.resolve(certPath, 'client.pem'),
    clientKey: path.resolve(certPath, 'client.key'),
    serverKey: path.resolve(certPath, 'server.key'),
    serverCert: path.resolve(certPath, 'server.pem'),
    caCert: path.resolve(certPath, 'ca.pem')
  },
  ssl: false,
  port: 3000,
  mongodb: {
    urlskel: 'mongodb://'+cred+'localhost:27017/%s?reconnect=true&authSource=admin',
    clientdb: 'pophealth-production',
    serverdb: 'notarization',
    logs: 'logs',
    nota: 'nota',
    since: new Date("2015-01-01")
  },
  notarizer:{
    host: "localhost",
    port: 3000,
    app: "notar",
    ok: "VALID Group"
  },
  zip:{
    active:true,
    dir: path.resolve(__dirname,'../archive'),
    maxFileSize: 100000000
  },
  https: {
    requestCert: true,
    rejectUnauthorized: false
  },
  routes: {
  },
  requestOpts:{
    passphrase:''
  },
  sha2: function(txt){
    return crypto.createHash('sha224').update(txt).digest('hex');
  }
};

function init(overrides) {
  config.zip.fileName = config.mongodb.clientdb+'-logs';
  config = extend(true, config, overrides);
  if (config.ssl) {
    config.clientCert = fs.readFileSync(config.certs.clientCert);
    config.clientKey = fs.readFileSync(config.certs.clientKey);
    config.serverCert = fs.readFileSync(config.certs.serverCert);
    config.serverKey = fs.readFileSync(config.certs.serverKey);

    config.caCert = fs.readFileSync(config.certs.caCert);
    config.https.key = config.serverKey;
    config.https.cert = config.serverCert;
    config.https.ca = config.caCert;
    config.requestOpts.cert = config.clientCert;
    config.requestOpts.key = config.clientKey;
    config.requestOpts.ca = config.caCert;
  }
  return config;
}

exports.init = init;


routes/notar-routes.js
// notar-routes.js
var path = require('path'),
    app = require(path.resolve(__dirname, '../app')),
    fs = require('fs'),
    bunyan = require('bunyan'),
    log = bunyan.createLogger({name: 'notar-routes', "level": "info"}),
    mongoskin = require('mongoskin'),
    extend = require('node.extend'),
    util = require('util'),
    Zipper = require('jszip'),
    configPath = path.resolve(__dirname, '../config/'),
    http = require('http'),
    https = require('https'),
    argv = require('yargs')
        .usage("node notar-routes.js [--port PORT] [ --ssl ] [--level LOGLEVEL] [--serverdb SERVER_DATABASE")
        .boolean('ssl')
        .number('port')
        .argv
    ;

var config;

var calls = {
  help: "/notar/help",
  sign: "POST /notar/sign {hash: <hash>}",
  validate: "POST /notar/validate {hash: <hash>, signature: <key>, time:<timestamp>",
  NOTE: "9180 requires authentication"
};
var coll; // db collection

function init() {
  app.get('/notar/help', function (req, res, next) {
    res.send(JSON.stringify(util.format(calls)));
  })

  app.post('/notar/sign', function (req, res, next) {
        var hash = req.body.hash;
        var ip = req.headers['x-forwarded-for'] || req.connection.remoteAddress;
        var id = req.body.id;
        var zipped = req.body.zipped;
        var newzip = new Zipper();
        if (!zipped) return saveIt();
        var zipin = new Zipper();
        zipin.loadAsync(zipped)
            .then(function (z) {
              saveIt(z)
            })
            .catch(function (err) {
              log.error(err, "loadasync zipped failed")
            })


        function saveIt(z) {
          if (!z) return signIt();

          function zipIt(newzip) {
            newzip.generateNodeStream({type: 'nodebuffer', streamFiles: true})
                .pipe(fs.createWriteStream(zipFilePath))
                .on('finish', function () {
                  signIt();
                });
          }

          var zipFilePath = path.resolve(config.zip.dir, config.zip.fileName + '.zip');
          if (!fs.existsSync(zipFilePath)) {
            // first time create zip file
            fs.mkdirSync(config.zip.dir);
            return zipIt(z);
          }
          var size = fs.lstatSync(zipFilePath).size;
          if (size >= config.zip.size) {
            var rs = fs.createReadStream(zipFilePath);
            var ws = fs.createWriteStream(zipFilePath + new Date().toString);
            rs.on('open', function (fd) {
              rs.pipe(ws);
            })
            rs.on('error', function (err) {
              log.error(err, "Read copying zip file to backup")
            })
            ws.on('error', function (err) {
              log.error(err, "Write copying zip file to backup")
            })
            ws.on('close', function () {
              zipIt(z); //it's just a new file
            })
          } else {
            // todo: jszip seems to have no way to add to a file on disk (like jar appears to do)
            var data = "";
            fs.createReadStream(zipFilePath, 'binary')
                .on('data', function (chunk) {
                  data += chunk;
                })
                .on('close', function () {
                  newzip.loadAsync(data)
                      .then(function (newzip) {
                        return newzip;
                      })
                      .then(function (newzip) {
                        return newzip.loadAsync(zipped)
                      })
                      .then(function (newzip) {
                        zipIt(newzip);
                      })
                      .catch(function (err) {
                        log.error(err, "Adding new file to zip");
                        // return to client
                        signIt();
                      })
                })
          }
        }

        function signIt() {
          var millis = new Date().valueOf();
          if (!hash) return res.status(400).send("/notar/sign: No hash element in post body");
          try {
            coll.insertOne({ip: ip, itemid: id, hash: hash, time: millis},// time: new Date(body.timestamp)},
                function (err, wresult) {
                  if (err) {
                    log.error(err, "signing hash from %s", ip);
                    return res.status(500).send("/notar/sign: Error inserting hash");
                  }
                  res.send(JSON.stringify({time: millis, signature: wresult.insertedId.toString()}));
                })
          } catch (e) {
            log.error(e, "writing db");
          }
        }
      }
  );

  var numberValidated = 0;
  // # count
  // Return count between dates and total in notarization collection
  app.get('/notar/count', function (req, res, next) {
    // this is the first step in validation.
    numberValidated = 0;
    var start = parseInt(req.query.start),
        end = parseInt(req.query.end);
    coll.count({}, function (err, total) {
      if (err) {
        log.error(err, "getting total count");
        return res.status(500).send("ERROR getting count.");
      }

      coll.count({$and: [{time: {$gte: start}}, {time: {$lte: end}}]}, function (err, count) {
        if (err) {
          log.error(err, "getting count");
          return res.status(500).send("ERROR getting count.");
        }
        res.send(JSON.stringify({count: count, total: total}));
      })
    })
  })

  app.post('/notar/validate', function (req, res, next) {
    var arr = req.body;
    numberValidated += arr.length;
    var invalid = [];
    doOne();
    function doOne() {
      if (!arr || !arr.length) {
        log.info("cumulative number validated: %d", numberValidated);
        if (invalid.length) {
          return res.send(JSON.stringify({failed: invalid}))
        } else {
          return res.send({valid: config.notarizer.ok, numberValidated: numberValidated});
        }
      }
      var item = arr.shift();
      var signature = item.signature, itemid = item.itemid, hash = item.hash;
      coll.findOne({_id: mongoskin.ObjectId(signature)}, function (err, ob) {
        if (err) {
          log.error(err, "finding doc with signature: %s", signature);
          return res.status(500).send("processing error. Please check notarizer db and retry.");
        }
        if (!ob) {
          log.error("No object for signature:%s", signature);
          return invalid.push({invalid: true, reason: "no such signature"});
        }
        if (ob.hash == hash && ob.itemid == itemid) return setImmediate(doOne);
        // errors
        if (ob.hash != hash) invalid.push({invalid: true, reason: "hash does not match", itemid: itemid})
        else if (ob.itemid != itemid) invalid.push(
            {invalid: true, reason: "log item id does not match", itemid: itemid})
        //else
        coll.update({_id: mongoskin.ObjectId(signature)}, {$set: {invalid: true}},
            function (err, up) {
              if (err) {
                log.error(err, "updating invalid id")
              }
              setImmediate(doOne);
            })
      })
    }
  })

  app.use(function (req, res, next) {
    var err = new Error('Not Found');
    log.error("Not found. url:%s, query:%j, body:%j", req.url, req.query, req.body);
    err.status = 404;
    next(err);
  });

}
//================  MAIN ======================
if (require.main == module) {
  var configFile = argv.config || path.resolve(configPath, 'config.js');
  config = require(configFile).init();
  if (argv.level) log.level(argv.level);
  if (argv.serverdb) config.mongodb.serverdb = argv.serverdb;
  if (argv.ssl) config.ssl = true;
  if (argv.port) config.port = argv.port;

  var db;
  try {
    db = mongoskin.MongoClient.connect(
        util.format(config.mongodb.urlskel, config.mongodb.serverdb),
        {native_parser: true});
  }
  catch (err) {
    dbConnectFailure(err);
  }

  coll = db.collection(config.mongodb.nota);

  function dbConnectFailure(err) {
    log.error('Unable to connect to mongo at %s.\n   ', config.mongodb.urlskel, err.message);
    setTimeout(function () {process.exit(1);}, 3000);
  }

  init();
  startServer();
  var server;

  function startServer() {
    if (config.ssl) server = https.createServer(config.https, app);
    else server = http.createServer(app);
    server.listen(config.port);
    console.log(new Date() + ": " + (config.ssl ? "HTTPS" : "HTTP") + " Listening at " + config.port);
  }

}
